### ISSUE TITLE
CA2007 false positive for any user-defined async disposable wrapper

### ISSUE THREAD

**Issue by saltukkos** [dotnet/roslyn-analyzers#6302]
### Analyzer

**Diagnostic ID**: [CA2007](https://learn.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/quality-rules/ca2007): `Do not directly await a Task`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: SDK 6.0.100

### Describe the bug

Analyzer triggers false-positive warning with `await using` of any user-defined wrapper of async disposable. 

### Steps To Reproduce

1. Define type like this 
```
    [StructLayout(LayoutKind.Auto)]
    public readonly struct 
        ConfiguredAsyncDisposable2
    {
        private readonly IAsyncDisposable _source;
        private readonly bool _continueOnCapturedContext;
        
        public ConfiguredAsyncDisposable2(IAsyncDisposable source, bool continueOnCapturedContext)
        {
            _source = source;
            _continueOnCapturedContext = continueOnCapturedContext;
        }
        
        public ConfiguredValueTaskAwaitable DisposeAsync() => _source.DisposeAsync().ConfigureAwait(_continueOnCapturedContext);
    }
```
2. Define corresponding extension method 
```
    public static class TaskAsyncEnumerableExtensions2
    {
        public static ConfiguredAsyncDisposable2 ConfigureAwait2(this IAsyncDisposable source, bool continueOnCapturedContext) =>
            new ConfiguredAsyncDisposable2(source, continueOnCapturedContext);
    }
```
3. Try to use it:
```
await using var foo = new FileStream("foo", FileMode.Open).ConfigureAwait2(false);
```

### Expected behavior
No warnings produced

### Actual behavior
[CA2007] Consider calling ConfigureAwait on the awaited task

## Why is it a bug

Short answer: there is no "ConfigureAwait" at all.

Long answer:
Original implementation of analyzer (before supporting `await using` syntax) handles this well. User can define their own awaitable types which can even don't have `.ConfigureAwait` extension methods. For that purpose in method TryGetTaskTypes analyzer collects target types, which are considered "bad" to be directly awaited https://github.com/dotnet/roslyn-analyzers/blob/628ee9a27c8ab7c2d648a6f2e59eba1eea623463/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs#L145 
However, in `async using` implementation only one type (WellKnownTypeNames.SystemRuntimeCompilerServicesConfiguredAsyncDisposable) is considered to be "good", and it's an unexpected behavior. 
I believe analyzer should check the return type of `DisposeAsync` method and use the same rules for it. (So it can be any type except the Task\Task<>\ValueTask\ValueTask<>).

**Comment by WalkerCodeRanger** [comment]
We also ran into this issue though our code is slightly different from the example code provided. We have a custom awaitable type `Result` and our own interface `IResultDisposable`.
```c#
public interface IResultDisposable
{
    Result DisposeAsync();
}
``` 
A `Result` does not need, nor have, a `ConfigureAwait` method.

CA2007 should either support this or at a minimum, the checking of this should be in a separate rule so we can disable it only there.

**Comment by DoctorKrolic** [comment]
This issue was fixed in https://github.com/dotnet/sdk/pull/55036. @tannergooding Can you please close this one as completed?
